### PR TITLE
Initialize all member variables in TrilinosWrappers::MPI::Vector

### PR DESCRIPTION
**Pull request by masterleinad**
After merging `TrilinosWrappers::Base` into `TrilinosWrappers::MPI::Vector` in #4454, Coverity reports some uninitialized variables. Since we are requiring C++11 now, we might as well use default member initializer.

### CODE REVIEW

**Comment by drwells** [comment]
I am slightly opposed to this on stylistic grounds: I don't think that we use this style of initialization anywhere else. I would rather use inheritance to set up default values in multiple constructors than something like this.

Does anyone else have an opinion on this?

**Comment by Rombur** [comment]
Personally, I like it. I don't have to think about initializing the variables when I add a new constructor but I can see why some don't like. Not a big deal for me though.

**Comment by bangerth** [comment]
I am not the biggest fan of this approach either because it makes it so difficult to distinguish the initializer here from the initializers of `const` variables. If this propagates to the doxygen page, for example, I would not at all be surprised if people start thinking that the variable has a constant value, rather than this just being an initializer.

**Comment by drwells** [comment]
Well, that is two for and four against. I think we can all agree that we don't want to write ten constructors that each start with
```
last_action(Zero),
compressed(true),
has_ghosts(false)
```

The best alternative I can come up with is something like this:
```cpp
namespace TrilinosWrappers
{
  /**
   * A class whose members describe the state (ghost status and last action) of
   * a Trilinos vector.
   */
  class VectorState
  {
  public:
    Epetra_CombineMode last_action;
    bool compressed;
    bool has_ghosts;

    VectorState();
  };

  VectorState::VectorState() :
    last_action(Zero),
    compressed(true),
    has_ghosts(false)
  {}

  [.......]

  class Vector : public Subscriptor, protected VectorState
  {
  [....]
  };
}

**Comment by bangerth** [comment]
I don't see anything particularly wrong with initializing the same variables the same way in multiple constructors. We could also use delegating constructors if you prefer that -- that would mean only one constructor has the initialization of the members in question, and all others delegate to it.

**Comment by masterleinad** [comment]
So be it. I will switch back to delegating constructors.

**Comment by masterleinad** [comment]
The only issue I see is that we always initialize `std::shared_ptr< Epetra_FEVector > vector` empty first. Hopefully, this is not too expensive.

**Comment by drwells** [comment]
Today I learned a new C++11 feature (delegating constructors). That is definitely the way to go.

File: source/lac/trilinos_vector.cc [hunk 1]
```diff
@@ -812,6 +799,7 @@ namespace TrilinosWrappers
       std::swap(last_action, v.last_action);
       std::swap(compressed, v.compressed);
       std::swap(vector, v.vector);
```
Comment (drwells) [inline source/lac/trilinos_vector.cc]: There are a few other members that are not swapped here, like `nonlocal_vector` and `owned_elements`. Should they be here?

**Comment by oneliefleft** [comment]
Today I learned a new C++11 feature (delegating constructors). That is
definitely the way to go.


Me too! Nice way forward...  :-)

File: source/lac/trilinos_vector.cc (hunk 1 above)
Comment (masterleinad) [inline source/lac/trilinos_vector.cc]: Yes, that makes sense.

**Comment by bangerth** [comment]
Glad I can still be of occasional value ;-)

**Pull request merged**